iceoryx_posh/test/integrationtests/test_service_discovery.cpp

```diff
@@ -473,31 +485,124 @@ TEST_F(ServiceDiscovery_test, FindServiceReturnsMaxServices)
+                     iox::popo::createNotificationCallback(testCallback))
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+
+    const iox::capro::ServiceDescription SERVICE_DESCRIPTION("Moep", "Fluepp", "Shoezzel");
+    iox::popo::UntypedPublisher publisher(SERVICE_DESCRIPTION);
+
+    auto notificationVector = waitSet.wait();
+
+    for (auto& notification : notificationVector)
     {
-        std::string instance = "i" + iox::cxx::convert::toString(i);
-        iox::capro::ServiceDescription SERVICE_DESCRIPTION(
-            SERVICE, IdString_t(iox::cxx::TruncateToCapacity, instance), "foo");
-        publishers.emplace_back(SERVICE_DESCRIPTION);
+        (*notification)();
     }
 
-    auto serviceContainer = sut.findService(SERVICE, iox::capro::Wildcard, iox::capro::Wildcard);
+    EXPECT_TRUE(callbackWasCalled);
+}
+
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryNotifiedbyWaitSetFindsSingleService)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1ecde7e0-f5b2-4721-b309-66f32f40a7bf");
+    iox::popo::WaitSet<1U> waitSet;
+    iox::capro::ServiceDescription serviceDescriptionToSearchFor("Soep", "Moemi", "Luela");
+
+    waitSet
+        .attachEvent(sut,
+                     ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED,
+                     0U,
+                     iox::popo::createNotificationCallback(searchForService, serviceDescriptionToSearchFor))
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to wait set"; });
+
+    iox::popo::UntypedPublisher publisher(serviceDescriptionToSearchFor);
+
+    auto notificationVector = waitSet.wait();
+
+    for (auto& notification : notificationVector)
+    {
+        (*notification)();
+    }
+
+    ASSERT_FALSE(serviceContainer.empty());
+    EXPECT_THAT(serviceContainer.front(), Eq(serviceDescriptionToSearchFor));
+}
+
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryIsAttachableToListener)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "def201f7-d1bf-4031-8e50-a2ad22ee303c");
+    iox::popo::Listener listener;
+
+    listener
+        .attachEvent(
+            sut, ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED, iox::popo::createNotificationCallback(testCallback))
+        .and_then([]() { GTEST_SUCCEED(); })
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to listener"; });
+}
+
+TEST_F(ServiceDiscovery_test, ServiceDiscoveryIsNotifiedByListenerAboutSingleService)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "305107fc-41dd-431c-8032-ed5e82f93038");
+    iox::popo::Listener listener;
+
+    listener
+        .attachEvent(
+            sut, ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED, iox::popo::createNotificationCallback(testCallback))
+        .or_else([](auto) { GTEST_FAIL() << "Could not attach to listener"; });
+
+    const iox::capro::ServiceDescription SERVICE_DESCRIPTION("Moep", "Fluepp", "Shoezzel");
+    iox::popo::UntypedPublisher publisher(SERVICE_DESCRIPTION);
+
+    this->InterOpWait();
```

Comment:
Please do not use `InterOpWait` anymore since it fails to often in our CI. Use the following approach:
1. include `watch_dog.hpp` and use it as a member in the test like:
```cpp
void SetUp() {
  m_watchdog.watchAndActOnFailure([]{ EXPECT_TRUE(false); std::terminate(); });
}

const iox::units::Duration m_fatalTimeout = 2_s;
Watchdog m_watchdog{m_fatalTimeout};
```
2. Introduce a `std::atomic_bool wasCallbackCalled{false};` in your test in the first line.
3. Set this atomic in the notification callback to true.
4. Introduce a semi busy loop to wait for it
```cpp
while(!wasCallbackCalled.load()) {
  std::this_thread::yield();
}
```
5. At the end you can write something like `EXPECT_TRUE(wasCallbackCalled.load());` The watchdog watches the system so that there is no deadlock occurring and this approach is valid. This will make run the test much faster and reliable and if this does not work the watchdog will terminate the unit test to prevent it from blocking the CI.


Please follow this approach for every instance where you used `InterOpWait`

Reply:
For old tests we may postpone this to #1065 

For new ones I agree.

Reply:
Good catch! Adapted for new test, existing tests to be done as part of #1065